esp-link/mqtt_client.c

```diff
@@ -3,91 +3,137 @@
+  if (connected_cb)
+    connected_cb(args);
 }
 
 void ICACHE_FLASH_ATTR
 mqttDisconnectedCb(uint32_t *args) {
 //  MQTT_Client* client = (MQTT_Client*)args;
-  os_printf("MQTT Disconnected\n");
-}
-
-void ICACHE_FLASH_ATTR
-mqttTcpDisconnectedCb(uint32_t *args) {
-//  MQTT_Client* client = (MQTT_Client*)args;
-  os_printf("MQTT TCP Disconnected\n");
+  DBG_MQTTCLIENT("MQTT Client: Disconnected\n");
+  if (disconnected_cb)
+    disconnected_cb(args);
 }
 
 void ICACHE_FLASH_ATTR
 mqttPublishedCb(uint32_t *args) {
 //  MQTT_Client* client = (MQTT_Client*)args;
-  os_printf("MQTT Published\n");
+  DBG_MQTTCLIENT("MQTT Client: Published\n");
+  if (published_cb)
+    published_cb(args);
 }
 
 void ICACHE_FLASH_ATTR
 mqttDataCb(uint32_t *args, const char* topic, uint32_t topic_len, const char *data, uint32_t data_len) {
+  //  MQTT_Client* client = (MQTT_Client*)args;
+
   char *topicBuf = (char*)os_zalloc(topic_len + 1);
   char *dataBuf = (char*)os_zalloc(data_len + 1);
 
-//  MQTT_Client* client = (MQTT_Client*)args;
-
   os_memcpy(topicBuf, topic, topic_len);
   topicBuf[topic_len] = 0;
 
   os_memcpy(dataBuf, data, data_len);
   dataBuf[data_len] = 0;
 
-  os_printf("Receive topic: %s, data: %s\n", topicBuf, dataBuf);
+  DBG_MQTTCLIENT("MQTT Client: Received topic: %s, data: %s\n", topicBuf, dataBuf);
   os_free(topicBuf);
   os_free(dataBuf);
+
+  if (data_cb)
+    data_cb(args, topic, topic_len, data, data_len);
 }
 
-  MQTT_InitConnection(&mqttClient, MQTT_HOST, MQTT_PORT, MQTT_SECURITY);
-  MQTT_InitClient(&mqttClient, MQTT_CLIENT_ID, MQTT_USER, MQTT_PASS, MQTT_KEEPALIVE, MQTT_CLSESSION);
-  MQTT_InitLWT(&mqttClient, "/lwt", "offline", 0, 0);
-  MQTT_OnConnected(&mqttClient, mqttConnectedCb);
-  MQTT_OnDisconnected(&mqttClient, mqttDisconnectedCb);
-  MQTT_OnDisconnected(&mqttClient, mqttTcpDisconnectedCb);
-  MQTT_OnPublished(&mqttClient, mqttPublishedCb);
-  MQTT_OnData(&mqttClient, mqttDataCb);
-#endif
+void ICACHE_FLASH_ATTR
+wifiStateChangeCb(uint8_t status)
+{
+  if (flashConfig.mqtt_enable) {
+    if (status == wifiGotIP  && mqttClient.connState != TCP_CONNECTING) {
+      MQTT_Connect(&mqttClient);
+    }
+    else if (status == wifiIsDisconnected && mqttClient.connState == TCP_CONNECTING) {
+      MQTT_Disconnect(&mqttClient);
+    }
+  }
+}
+
+void ICACHE_FLASH_ATTR
+mqtt_client_init()
+{
+  if (flashConfig.mqtt_enable) {
+    MQTT_Init(&mqttClient, flashConfig.mqtt_host, flashConfig.mqtt_port, 0, flashConfig.mqtt_timeout,
+      flashConfig.mqtt_clientid, flashConfig.mqtt_username, flashConfig.mqtt_password,
+      flashConfig.mqtt_keepalive);
+
+    if (flashConfig.mqtt_status_enable) {
+      statusTopicStr = (char*)os_zalloc(strlen(flashConfig.mqtt_clientid) + strlen(flashConfig.mqtt_status_topic) + 2);
+      os_strcpy(statusTopicStr, flashConfig.mqtt_clientid);
+      os_strcat(statusTopicStr, "/");
+      os_strcat(statusTopicStr, flashConfig.mqtt_status_topic);
+    }
```

Comment:
this creates an inconsistency between what the user sees on the web page and what is actually used, and while you may like the clientid to be prepended that's not necessarily other's preference.